packages/amplify-graphql-api-construct/src/amplify-graphql-definition.ts

```diff
@@ -1,32 +1,90 @@
 import * as os from 'os';
 import { SchemaFile } from 'aws-cdk-lib/aws-appsync';
-import { IAmplifyGraphqlDefinition } from './types';
+import { IAmplifyGraphqlDefinition, ModelDataSourceDefinition } from './types';
+import { getModelTypeNames } from './internal';
 
+export const DEFAULT_MODEL_DATA_SOURCE_DEFINITION: ModelDataSourceDefinition = {
+  name: 'defaultDDB',
+  strategy: {
+    dbType: 'DYNAMODB',
+    provisionStrategy: 'DEFAULT',
+  },
+};
 /**
  * Class exposing utilities to produce IAmplifyGraphqlDefinition objects given various inputs.
  */
 export class AmplifyGraphqlDefinition {
   /**
    * Produce a schema definition from a string input
    * @param schema the graphql input as a string
+   * @param modelDataSourceDefinition the provision definition for `@model` datasource. The DynamoDB from CloudFormation will be used by default.
    * @returns a fully formed amplify graphql definition
    */
-  static fromString(schema: string): IAmplifyGraphqlDefinition {
+  static fromString(
+    schema: string,
+    modelDataSourceDefinition: ModelDataSourceDefinition = DEFAULT_MODEL_DATA_SOURCE_DEFINITION,
+  ): IAmplifyGraphqlDefinition {
     return {
       schema,
       functionSlots: [],
+      dataSourceProvisionConfig: {
+        project: modelDataSourceDefinition,
+      },
     };
   }
 
   /**
    * Convert one or more appsync SchemaFile objects into an Amplify Graphql Schema
    * @param filePaths one or more paths to the graphql files to process
+   * @param modelDataSourceDefinition the provision definition for `@model` datasource. The DynamoDB from CloudFormation will be used by default.
    * @returns a fully formed amplify graphql definition
    */
-  static fromFiles(...filePaths: string[]): IAmplifyGraphqlDefinition {
+  static fromFiles(
+    filePaths: string | string[],
+    modelDataSourceDefinition: ModelDataSourceDefinition = DEFAULT_MODEL_DATA_SOURCE_DEFINITION,
+  ): IAmplifyGraphqlDefinition {
+    if (!Array.isArray(filePaths)) {
+      filePaths = [filePaths];
+    }
     return {
       schema: filePaths.map((filePath) => new SchemaFile({ filePath }).definition).join(os.EOL),
       functionSlots: [],
+      dataSourceProvisionConfig: {
+        project: modelDataSourceDefinition,
+      },
+    };
+  }
+
+  /**
+   * Combines multiple IAmplifyGraphqlDefinitions into a single definition.
+   * @param definitions the definitions to combine
+   */
+  static combine(...definitions: IAmplifyGraphqlDefinition[]): IAmplifyGraphqlDefinition {
+    if (definitions.length === 0) {
+      throw new Error('The definitions of amplify GraphQL cannot be empty.');
+    }
+    if (definitions.length === 1) {
+      return definitions[0];
+    }
+    const projectStrategy = definitions[0].dataSourceProvisionConfig.project;
```

Comment:
It looks like you're using the `project` config of the first definition to apply to all subsequent definitions, but that's not how it should work. Each definition should be able to supply its own configuration, which will be independently resolved. If we really want a top-level config (which I would argue against), then it should be a separate argument to the `combine` method.

Reply:
The first definition is not applied to all subsequent definitions unless the latter strategy is different from the first one(either `name` or `provisionStrategy`, or other fields related to RDS usage). When the difference is detected, the corresponding config will be keyed by the model type name.

Reply:
The initial thought on this is that if the user only uses one strategy (like the default DDB), we don't need to map every model type name to the same value but only use a default value for all.

Reply:
But if I'm reading this right, you're losing all of the model-specific overrides in def 1 since you `slice()` the definition array but never store the explicitly mapped models?

E.g. in this code, you'd lose the override on the T2 model:

```typescript
const def1 = AmplifyGraphQlApiDefinition.fromString('type T1 @model {id: ID!} type T2 @model {id: ID!}', {
  models: {
    T2: {
      dbType: 'DYNAMODB',
      provisioningStrategy: 'AMPLIFY_MANAGED'
    }
  }
});

const def2 = AmplifyGraphQlApiDefinition.fromString('type T3 @model {id: ID!} type T4 @model {id: ID!}', {
  default: {
      dbType: 'DYNAMODB',
      provisioningStrategy: 'AMPLIFY_MANAGED'
  },
  models: {
    T4: {
      dbType: 'DYNAMODB',
      provisioningStrategy: 'DEFAULT'
    }
  }
});
```

Further, our internal implementation is going to be simpler if we can assume at every level other than the UI that there is an explicit mapping of model to data source. The factory methods, including `combine`, should just go through each definition and create an explicit map of model to ModelDataSourceDefinition, and pass that through the rest of the system. That's also the place where we can quickly detect model conflicts (e.g., if def2 above had attempted to redefine the `T1` model), allowing us to fail fast.